File: Second Seminar/SecondSeminar/src/main/java/com/example/SecondSeminar/post/application/PostServiceV2.java
package com.example.SecondSeminar.post.application;

import com.example.SecondSeminar.external.S3Service;
import com.example.SecondSeminar.external.exception.S3Exception;
import com.example.SecondSeminar.external.exception.S3ExceptionType;
import com.example.SecondSeminar.member.domain.Member;
import com.example.SecondSeminar.member.domain.MemberJpaRepository;
import com.example.SecondSeminar.post.domain.Post;
import com.example.SecondSeminar.post.domain.PostJpaRepository;
import com.example.SecondSeminar.post.dto.request.PostCreateRequest;
import java.io.IOException;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.multipart.MultipartFile;

@RequiredArgsConstructor
@Transactional(readOnly = true)
@Service
public class PostServiceV2 {
    private static final String POST_IMAGE_FOLDER_NAME = "posts/";

    private final MemberJpaRepository memberJpaRepository;
    private final PostJpaRepository postJpaRepository;
    private final S3Service s3Service;

    @Transactional
    public String createV2(PostCreateRequest request, MultipartFile image, Long memberId) {
        try {
            final String imageUrl = s3Service.uploadImage(POST_IMAGE_FOLDER_NAME, image);

            Member member = memberJpaRepository.findByIdOrThrowException(memberId);

            Post post = postJpaRepository.save(
                    Post.builderWithImageUrl()
                            .title(request.title())
                            .content(request.content())
                            .imageUrl(imageUrl)
                            .member(member)
                            .build()
            );

            return post.getId().toString();
        } catch (RuntimeException e) {
            throw new RuntimeException(e.getMessage());
        } catch (IOException e) {
            throw new S3Exception(S3ExceptionType.UPLOAD_FAILED);
        }

    }
}

File: Second Seminar/SecondSeminarKotlin/src/main/kotlin/com/example/SecondSeminarKotlin/post/application/PostServiceV2.kt
package com.example.SecondSeminarKotlin.post.application

import com.example.SecondSeminarKotlin.exception.BaseCustomException
import com.example.SecondSeminarKotlin.external.S3Service
import com.example.SecondSeminarKotlin.member.application.MemberService
import com.example.SecondSeminarKotlin.post.domain.Post
import com.example.SecondSeminarKotlin.post.domain.PostJpaRepository
import com.example.SecondSeminarKotlin.post.dto.request.PostCreateRequest
import com.example.SecondSeminarKotlin.post.exception.PostExceptionType
import org.springframework.data.repository.findByIdOrNull
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import org.springframework.web.multipart.MultipartFile

private const val POST_IMAGE_FOLDER_NAME = "posts/"

@Transactional(readOnly = true)
@Service
class PostServiceV2(
    private val memberService: MemberService,
    private val postJpaRepository: PostJpaRepository,
    private val s3Service: S3Service,
) {

    @Transactional
    fun createV2(request: PostCreateRequest, image: MultipartFile, memberId: Long): String {
        val imageUrl = s3Service.uploadImage(POST_IMAGE_FOLDER_NAME, image)

        val member = memberService.findMemberById(memberId)

        val post = postJpaRepository.save(
            Post(
                title = request.title,
                content = request.content,
                imageUrl = imageUrl,
                member = member
            )
        )

        return post.id.toString()
    }

    @Transactional
    fun deleteByIdV2(postId: Long) {
        val post: Post =
            postJpaRepository.findByIdOrNull(postId) ?: throw BaseCustomException(PostExceptionType.NOT_FOUND_POST)

        // post에 imageUrl 정보가 있을 때만 이미지 삭제
        if (!post.imageUrl.isNullOrBlank())
            s3Service.deleteImage(post.imageUrl!!)

        postJpaRepository.deleteById(postId)
    }
}